Step5 - 리펙터링

oyj7677 opened the pull request:
MVP패턴 구현
Presenter 유닛 테스트 구현

Commit:
feat: 사용자가 피연산자 0 ~ 9 버튼을 누르면 화면에 해당 숫자를 보여준다

Commit:
Merge branch 'main' of https://github.com/oyj7677/android-calculator-mvp

Commit:
Merge branch 'main' of https://github.com/oyj7677/android-calculator-mvp

Commit:
Merge remote-tracking branch 'upstream/oyj7677'
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

Commit:
step 4 / 2차 개선사항
- CalculatorLib -> domain 모듈 명칭 변경
- Formula 모듈 domain 모듈 일부 클래스로 흡수
- 수식 검증 유닛 테스트 진행

Commit:
Step5. 리팩터링
 - MVP패턴 적용
 - Presenter 단위 테스트 생성 - Presenter 단위테스트 설명 참고
 - 계산 결과 표시 수정 (결과 -> 수식\n= 결과)

oyj7677 commented:
2주차 강의를 보며 mock 테스트를 진행하고 있습니다.
아래의 코드에서 다음과 같은 에러가 발생하고 있습니다.
/**
no answer found for View(#2).showFormula(1) among the configured answers: ()
io.mockk.MockKException: no answer found for View(#2).showFormula(1) among the configured answers: ()
*/
해당 코드는 mock을 통해 view를 생성했지만 view의 함수는 구현하지 않아 생긴 에러라고 강의에서 설명해 주셨습니다.
그리하여 given 부분에 every { view.showFormula(any()) } just Runs를 작성하여 테스트 코드에서 view.showFormula()의 처리를 
단순하게 넘기도록 설정해 준 것으로 강의에서 말씀해 주셨습니다. 하지만 하단 코드의 경우 말씀해주신 것처럼 구현된 것으로 보이나 
여전히 위의 에러가 나오고 있습니다. 어느부분이 잘못됐는지 조언을 주시면 감사하겠습니다.

@Test
    fun `1 입력`() {
        // given :
        every { view.showFormula(any()) } just Runs

        // when :
        presenter.addNumber('1')
        presenter.addNumber('1')

        // then :
        val expected = "11"
        verify { view.showFormula(expected) }
    }

malibinYun commented on app/src/main/java/camp/nextstep/edu/calculator/Presenter.kt:
```diff
@@ -0,0 +1,46 @@
+    private val formula = Formula()
+    private val calculate = Calculator()
+
+    override fun getFormula(): String {
+       return formula.getFormula()
+    }
```
사용하지 않는 메서드는 제거해보는 건 어떨까요?

malibinYun commented on app/src/main/java/camp/nextstep/edu/calculator/Presenter.kt:
```diff
@@ -0,0 +1,46 @@
+    // 직전 값이 숫자인가
+    // return True: 숫자, False: 연산자
+    // 직전 값이 없다면?
+    override fun isLastStrNum(): Boolean {
+        return formula.isLastStrNum()
+    }
```
이 역할은 Presenter내부에서만 사용하네요!
public으로 공개될 필요가 없어 보입니다. 공개될 필요가 없으니, Presenter Interface에도 제거해보는 건 어떨까요?

malibinYun commented on app/src/main/java/camp/nextstep/edu/calculator/Presenter.kt:
```diff
@@ -0,0 +1,46 @@
+
+    // 수식 계산
+    override fun calculate() {
+        if (isLastStrNum()) {
+            view.showResult(calculate.evaluate(formula.getFormula()).toString())
```
메서드를 한 줄로만 작성하는 것이 깔끔해보일 수는 있지만, 가독성을 챙기기 어려울 때가 종종 있습니다.
또, 41번째 라인의 메서드 중 하나에서 Exception이 발생하더라도,
콘솔창에서는 41번째 라인에서 에러가 발생했다고 안내되어 디버깅이 어려워질 수 있게됩니다.
의미 단위로 지역 변수에 담아보는 건 어떨까요?

malibinYun commented on app/src/test/java/camp/nextstep/edu/calculator/DomainTest.kt:
```diff
@@ -0,0 +1,317 @@
+        assertThat(exception).isInstanceOf(IllegalArgumentException::class.java)
+    }
+
+    private var numAndOperator = "12 +"
+
+    @Test
+    fun 마지막_값을_검증한다() {
+        // given: 문자열의 마지막 인덱스를 구한다.
+        val lastIndex = numAndOperator.length - 1
+
+        // when: 문자열의 마지막 값이 숫자인지 확인합니다.
+        val actual = numAndOperator[lastIndex].isDigit()
+
+        // then: 숫자일 경우 true, 숫자 이외의 것 이라면 false
+        assertThat(actual).isEqualTo(false)
+    }
+
+    @Test
+    fun 마지막으로_입력된_연산자를_교체한다() {
+        // given: 문자열의 마지막 인덱스를 구한다.
+        val lastIndex = numAndOperator.length - 1
+
+        // when: 마지막 문자열의 값을 삭제 후 추가한다.
+        numAndOperator = numAndOperator.dropLast(1)
+        numAndOperator += "*"
+        val actual = numAndOperator[lastIndex]
+
+        // then: 숫자일 경우 true, 숫자 이외의 것 이라면 false
+        assertThat(actual).isEqualTo('*')
+    }
```
https://github.com/next-step/android-calculator-mvp/pull/210#discussion_r1285040516
여전히 유효한 피드백입니다.
이 테스트는 Calculator의 테스트가 맞는지 충분히 고민해보시면 좋겠어요 :)

malibinYun commented on app/src/test/java/camp/nextstep/edu/calculator/DomainTest.kt:
```diff
@@ -0,0 +1,317 @@
+    }
+}
+
+class FormulaTest {
+    var formula: Formula = Formula()
```
테스트 코드는 원본 코드가 있는 패키지 경로와 동일하게 맞추는 것이 컨벤션입니다.
또, 파일 이름은 "원본 코드의 파일 이름 + Test" 이 컨벤션입니다.

파일을 분리해보는 건 어떨까요?

malibinYun replied:
한 전역변수에 모든 테스트 메서드의 의존성이 강합니다.
이 전역 변수가 다른 사람이 추가한 테스트 코드에 의해서 다른 테스트 코드가 동작하지 않게 될 수도 있는 구조입니다.
한 테스트 메서드는 독립적인 하나의 프로그램이라고 생각하시는 게 가장 편합니다 :)
테스트 메서드 간에 아무런 의존성이 없도록 만들어보는 건 어떨까요?

malibinYun commented on app/src/test/java/camp/nextstep/edu/calculator/MainActivityLogicTest.kt:
```diff
@@ -1,35 +1,5 @@
-import com.google.common.truth.Truth.assertThat
-import org.junit.Test
-
 class MainActivityLogicTest {
 
```
필요없게된 파일은 삭제하는 건 어떨까요?

malibinYun commented on app/src/test/java/camp/nextstep/edu/calculator/PresenterTest.kt:
```diff
@@ -0,0 +1,149 @@
+        // then : 수식의 결과값을 보여줍니다.
+        assertThat(actual).isEqualTo("8")
+    }
+
+    // region Fixture
+    // 가짜 뷰
+    private fun createView(onShow: (String) -> Unit): Contract.View = object : Contract.View {
```
가짜 뷰를 통해 테스트를 잘 작성해주셨네요 👍

showResult가 불리는 것은 onShow를 통해 검증해서는 안됩니다. Presenter에 명령한 기능으로 showFormula, showResult 등이 동시에 호출될 수 있는 기획 변경 상황이 생길 수 있기 때문이에요.

View의 모킹은 모두 각각 해주어야합니다.

showToast가 불리는 상황도 테스트해보는 건 어떨까요?

malibinYun commented on domain/src/main/java/com/example/calculatorlib/Formula.kt:
```diff
@@ -0,0 +1,63 @@
+
+    fun getFormula() = formula
+
+    // 테스트를 위한 setter
+    fun setFormula(formula: String) {
+        this.formula = formula
+    }
```
테스트만을 위한 메서드는 프로덕션 코드에서 이 메서드를 오용하기 쉽게 만듭니다.
따로 스트만을 위한 메서드를 만들지 않고 해당 객체를 테스트 가능하게 만들어보는 건 어떨까요?

원하는 String 수식을 생성자로부터 초기화할 수 있게 만들어보는 건 어떨까요?

malibinYun reviewed (changes requested):
5단계 미션 고생 많으셨습니다.
이번이 마지막 미션인데요!
더 반영하고 싶으신 내용이 있으시다면 반영하시고 리뷰 요청 부탁드리고,
이만 마무리 짓고 싶으시다면 DM 혹은 코멘트 남겨주시면 미션 마무리 하겠습니다 :)

malibinYun commented:
> 2주차 강의를 보며 mock 테스트를 진행하고 있습니다.
아래의 코드에서 다음과 같은 에러가 발생하고 있습니다.

mockk로 view를 어떻게 생성하고 있는지 알 수 있을까요?
그리고 만들어진 View와 Presenter가 서로 참조할 수 있게 엮여있는지도 첨부해주신 코드로는 확인할 수 없어서,
에러 메시지만으로는 정확한 원인을 알기 어렵네요 😥

Commit:
Step5 개선
- 사용하지 않는 메서드 삭제
- Presenter내부에서만 사용하는 메서드 Presenter Interface에서 삭제
- 한줄로 여러 메서드가 사용된 코드 수정
- test코드 수정

oyj7677 commented:
테스트 클래스의 컨벤션에대한 질문이 있습니다.
이전 직전 리뷰에서 원본 코드가 있는 패키지 경로와 테스트 코드의 패키지 경로를 맞추는 것이 컨벤션이며
그로인해  FormulaTest 클래스를 분리해보는 것을 권장해주셨습니다.

이에 DomainTest.kt안에 있던 FormulaTest 클래스를 새로운 클래스파일로 분리하였습니다.

여기서 저의 궁금증은 그렇다면 DomainTest.kt에는 Calculator에 관한 내용만이 들어있게 되기 때문에 DomainTest.kt가 아니라 CalculatorTest.kt가 되지 않을까 생각이 들었습니다. 

이러한 질문의 이유는 일전에 CalculatorLib와 FormulaLib를 DomainLib 하나로 합치는 과정이 있었습니다.
 DomainLib에 대한 테스트 클래스를 DomainTest.kt라는 하나의 클래스를 만들어 Calculator와 Formula에 관련한 테스트를 모두 진행하였습니다.
이후 Formula에 대한 테스트를 분리하라는 조언에 DomainTest의 이름도 변경해야하지 않을까 생각이 들었습니다.

Commit:
도메인 라이브러리 패키지명 변경
com.example.calculatorlib -> com.example.domainlib

도메인 라이브러리 관련 유닛 테스트 패키지 생성 및 기본 클래스 이동

oyj7677 commented:
테스트 클래스 경로와 관련하여 패키지 수정을 일부 해보았습니다.
제대로 이해한 것인지 확인해주시면 감사하겠습니다.

malibinYun commented on app/src/test/java/camp/nextstep/edu/calculator/MainActivityTest.kt:
```diff
@@ -0,0 +1,5 @@
+
+import org.junit.Assert.*
+
+class MainActivityTest
```
MainActivity에 대한 테스트는 이미 androidTest 패키지에 존재하니, 유닛 테스트를 위한 곳에 꼭 추가하지 않으셔도 좋습니다 :)

malibinYun commented on app/src/test/java/camp/nextstep/edu/calculator/PresenterMockTest.kt:
```diff
@@ -0,0 +1,126 @@
+import io.mockk.verify
+import org.junit.Test
+
+class PresenterMockTest {
+    private val view: Contract.View = mockk(relaxed = true)
+    private val presenter = Presenter(view)
```
질문해주셨던 mockk의 에러에 대해 해결하셨나보네요! 👍👍

malibinYun commented on app/src/test/java/camp/nextstep/edu/calculator/PresenterTest.kt:
```diff
@@ -128,22 +142,43 @@ class PresenterTest {
     // region Fixture
     // 가짜 뷰
-    private fun createView(onShow: (String) -> Unit): Contract.View = object : Contract.View {
+    private fun createView(onShowFormula: (String) -> Unit,
+                           onShowResult: (String) -> Unit,
+                           onShowToast: (String) -> Unit): Contract.View = object : Contract.View {
```
디폴트 파라미터를 활용하면, 테스트에 꼭 필요하지 않은 람다를 불필요하게 넘기지 않게 줄여볼 수 있습니다 :)

malibinYun commented on app/src/test/java/com/example/domainlib/FormulaTest.kt:
```diff
@@ -0,0 +1,153 @@
+        formula.addNumber('1')
+
+        // then : 문자열에 추가된다.
+        Truth.assertThat(formula.getFormula()).isEqualTo("1")
```
static import를 하시면, `Truth.` 작성 없이, 바로 assertThat()~ 으로 코드를 작성할 수 있습니다 :)

malibinYun commented on app/src/test/java/com/example/domainlib/CalculatorTest.kt:
```diff
@@ -1,14 +1,13 @@
-package camp.nextstep.edu.calculator
+package com.example.domainlib
```
>여기서 저의 궁금증은 그렇다면 DomainTest.kt에는 Calculator에 관한 내용만이 들어있게 되기 때문에 DomainTest.kt가 아니라 CalculatorTest.kt가 되지 않을까 생각이 들었습니다.
>
>이러한 질문의 이유는 일전에 CalculatorLib와 FormulaLib를 DomainLib 하나로 합치는 과정이 있었습니다.
DomainLib에 대한 테스트 클래스를 DomainTest.kt라는 하나의 클래스를 만들어 Calculator와 Formula에 관련한 테스트를 모두 진행하였습니다.
이후 Formula에 대한 테스트를 분리하라는 조언에 DomainTest의 이름도 변경해야하지 않을까 생각이 들었습니다.

네! 의견 주신 것에 저도 동의합니다 :)
도메인 모듈의 Calculator, Formula, Operator 모두 테스트 파일명으로 각각 CalculatorTest, FormulaTest, OperatorTest로 대응되어야 합니다.

다만, 현재 테스트 파일의 모듈 위치가 잘못되었는데요,
도메인 모듈에 대한 테스트 파일이 현재 도메인 모듈이 아닌, app 모듈에 위치하고 있네요!

도메인 모듈에 대한 테스트는 도메인 모듈에 위치하게 옮겨주시면 됩니다 🙂

또, 클래스 이름을 바꿀 때는 파일 명도 같이 바뀌었는지 확인해보는 습관을 들이면 좋습니다 :)

malibinYun commented on domain/src/main/java/com/example/domainlib/Formula.kt:
```diff
@@ -0,0 +1,57 @@
+
+    fun getFormula() = formula
+
+    // 숫자 추가
+    fun addNumber(char: Char): String {
+        if (formula.isEmpty()) {
```
수식을 String으로 관리하다보면, 어떤 값이 들어가있는지 예측하기가 쉽지 않아지는데요,
피연산자도 Operator처럼 객체로 만들어보는 건 어떨까요?
이 Fomula 객체에서는 String으로 연산자와 피연산자를 저장하는 것이 아니라, 이 수식에 필요한 요소들을 하나의 List로 관리하게 만들어볼 수 있습니다 :)

미션이 끝났더라도 도전해보시면 좋겠네요!

malibinYun reviewed (approved):
피드백 반영하시느라 고생 많으셨어요!
미션은 이만 마무리하겠습니다 :)
질문 주신 부분들은 피드백에 남겨두었습니다.
미션 진행하시면서 고민했던 것들이 앞으로 개발하실 때 많은 도움이 되셨으면 좋겠어요!
